libraries/libfc/include/fc/crypto/bls_signature.hpp

```diff
@@ -1,51 +1,136 @@
-#include <fc/reflect/reflect.hpp>
 #include <fc/reflect/variant.hpp>
+#include <fc/exception/exception.hpp>
 #include <bls12-381/bls12-381.hpp>
 
+
 namespace fc::crypto::blslib {
 
    namespace config {
       const std::string bls_signature_prefix = "SIG_BLS_";
    };
-   
-   class bls_signature
-   {
-      public:
 
-         bls_signature() = default;
-         bls_signature( bls_signature&& ) = default;
-         bls_signature( const bls_signature& ) = default;
-         explicit bls_signature( const bls12_381::g2& sig ){_sig = sig;}
+   // Immutable after construction.
+   // Provides an efficient wrapper around bls12_381::g2.
+   // Serialization form:
+   //   Non-Montgomery form and little-endian encoding for the field elements.
+   //   Affine form for the group element (the z component is 1 and not included in the serialization).
+   //   Binary serialization encodes x component first followed by y component.
+   class bls_signature {
+   public:
+      bls_signature() = default;
+      bls_signature(bls_signature&&) = default;
+      bls_signature(const bls_signature&) = default;
+      bls_signature& operator=(const bls_signature&) = default;
+      bls_signature& operator=(bls_signature&&) = default;
+
+      // throws if unable to convert to valid bls12_381::g2
+      explicit bls_signature(std::span<const uint8_t, 192> affine_non_montgomery_le);
+
+      // affine non-montgomery base64url with bls_signature_prefix
+      explicit bls_signature(const std::string& base64urlstr);
+
+      // affine non-montgomery base64url with bls_signature_prefix
+      std::string to_string() const;
+
+      const bls12_381::g2&            jacobian_montgomery_le() const { return _jacobian_montgomery_le; }
+      const std::array<uint8_t, 192>& affine_non_montgomery_le() const { return _affine_non_montgomery_le; }
 
-         // affine non-montgomery base64url with bls_signature_prefix
-         explicit bls_signature(const std::string& base64urlstr);
+      bool equal(const bls_signature& sig) const {
+         return _jacobian_montgomery_le.equal(sig._jacobian_montgomery_le);
+      }
+
+      template<typename T>
+      friend T& operator<<(T& ds, const bls_signature& sig) {
+         ds.write(reinterpret_cast<const char*>(sig._affine_non_montgomery_le.data()), sig._affine_non_montgomery_le.size()*sizeof(uint8_t));
+         return ds;
+      }
+
+      // Could use FC_REFLECT, but to make it obvious serialization matches bls_aggregate_signature implement via operator
+      template<typename T>
+      friend T& operator>>(T& ds, bls_signature& sig) {
+         ds.read(reinterpret_cast<char*>(sig._affine_non_montgomery_le.data()), sig._affine_non_montgomery_le.size()*sizeof(uint8_t));
+         constexpr bool check = true;  // check if invalid
+         constexpr bool raw   = false; // non-montgomery
+         std::optional<bls12_381::g2> g2 = bls12_381::g2::fromAffineBytesLE(sig._affine_non_montgomery_le, check, raw);

```

Comment:
I wish `fromAffineBytesLE` took `struct { bool check; bool raw; };` as parameter so we could call with:

`bls12_381::g2::fromAffineBytesLE(sig._affine_non_montgomery_le, { .check = true, .raw = false });